glasswall/utils.py

```diff
@@ -249,7 +249,104 @@ def get_library(library: str, directory: str):
     raise FileNotFoundError(f'Could not find any files: "{library_file_names}" under directory: "{directory}"')
```

Comment:
Are the files run through linters. I ran utils.py through _pylint_ using default options and got the attached diagnostics as attached:
[utils.py-PYLINT.txt](https://github.com/gw-engineering/glasswall-python-wrapper/files/12181894/utils.py-PYLINT.txt)

Reply:
Yes, [using flake8](https://github.com/gw-engineering/glasswall-python-wrapper/actions/runs/5670374565/job/15365139532?pr=70#step:5:6) with this `setup.cfg`:

```cfg
[flake8]
# ignore "E501 Line too long"
# ignore "F401 Module imported but unused" and "E402 Module level import not at top of file" in __init__.py
ignore=E501
per-file-ignores=__init__.py:F401,E402
```

Reply:
Why so many diagnostics when I ran through pylint? Are the issues being addressed?

Reply:
I think flake8 is more sensible in what it considers to be an issue, pylint seems overly strict.

Reply:
Well, that's both a good thing and a bad thing. Bear in mind I ran _pylint_ using its defaults so there is probably a way of getting at the more "interesting" diagnostics and filtering out the ones that are deemed over-zealous.

I think these are beyond cosmetic, though:
    Unnecessary "elif" after "return"
    Redefining built-in 'type'
    Catching too general exception Exception
    Access to a protected member _OPERATING_SYSTEM of a client class
    Unnecessary use of a comprehension, use dict(sorted(dict_.items())) instead.

Reply:
I actually think all of those are still cosmetic.

Reply:
My thinking was along the lines of this:
*  Unnecessary "elif" after "return" - Could this be showing a logic error in that area. Why has an _elif_ been coded after a _return_?
* Redefining built-in 'type' - Redefining a built-in type seems a dangerous thing to do, or is it common in Python, and scoped?
* Catching too general exception Exception - I was under the impression catching _Exception_ is not best practice.
* Access to a protected member OPERATING_SYSTEM of a client class - Presumably a member is protected for a reason, and accessing it sounds inadvisable.
* Unnecessary use of a comprehension, use dict(sorted(dict.items())) instead - This looks like it might be an efficiency issue. I'm sure we don't want to use process cycles unnecessarily.

Reply:
* Unnecessary "elif" after "return" is referring to this code:
```py
    if isinstance(file_, bytes):
        return file_
    elif isinstance(file_, bytearray):
        return bytes(file_)
    elif isinstance(file_, io.BytesIO):
        return file_.read()
    else:
        raise TypeError(file_)
```
and wants it instead to be written as:
```py
    if isinstance(file_, bytes):
        return file_
    if isinstance(file_, bytearray):
        return bytes(file_)
    if isinstance(file_, io.BytesIO):
        return file_.read()
    raise TypeError(file_)
```

both are functionally the same.


* Redefining built-in 'type'
Irrelevant, these args are mandatory for __exit__ but not used
```py
    def __exit__(self, type, value, traceback):
        os.chdir(self.old_cwd)
```


* Catching too general exception 
This is one that I can sort of agree with in some cases, mostly when there's multiple lines of code in the "try" block. If there's only one line I think it's very reasonable to `except Exception:`, even more so when it isn't immediately clear what exceptions could be raised, for example in the below code pylint gives a warning. We want to try to preload a dependency, and if it can't preload it we raise an error only if "ignore_errors" is "False". We don't care what the exception is, whether the dependency file path didn't exist, or the file is corrupt and couldn't be loaded, so catching "Exception" is okay.

```py
    for dependency in dependencies:
        # Try to load dependencies that exist
        if dependency not in missing_dependencies:
            try:
                ct.cdll.LoadLibrary(dependency)
            except Exception:
                if ignore_errors:
                    pass
                else:
                    raise
```

Interestingly the pass in the above code seems a bit redundant but was not picked up by pylint, and could instead be:
```py
            try:
                ct.cdll.LoadLibrary(dependency)
            except Exception:
                if not ignore_errors:
                    raise
```

* Access to a protected member _OPERATING_SYSTEM of a client class
I don't fully understand this warning, _OPERATING_SYSTEM isn't within any class - it's intended to be a constant used internally within the glasswall package. In hindsight these maybe should have just been stylised as constants in all uppercase without an underscore prefix. I'm hesitant to change their names now that they are being used elsewhere, such as in the GCI.

* Unnecessary use of a comprehension, use dict(sorted(dict.items())) instead
You're right, that is a performance increase. I'm going to maintain that it's outside of the scope of this PR though.

Reply:
I think the `elif` stack is better and better documents that a series of choices are being made, but I understand _pylint_'s diagnostic.

I think the performance one is definitely a good catch and worthy of a Work Item to be addressed at some time.

At worst, _pylint_ has drawn attention to certain things, and we can decide what's useful or not. I suppose most linters use PEP as guidelines where they can as well.